index.html

```diff
@@ -1280,6 +1270,158 @@ <h3>
+              "dir": "ltr",
+              "shortcuts": [{
+                "description": "Change language",
+                "description_localized": {
+                  "ar": { "lang": "en", "dir": "ltr", "value": "Change language" }
+                },
+                "icons": [{
+                  "src": "/icons/play-later.svg",
+                  "type": "image/svg+xml"
+                }],
+                "icons_localized": {
+                  "ar": [{
+                    "src": "/icons/ar/play-later.svg",
+                    "type": "image/svg+xml"
+                  }]
+                }
+              }]
+            }
+          </pre>
+        </aside>
+        <p>
+          <!-- todo: is there an "any" type? -->
+           To <dfn>normalize the localized value</dfn>, given |localizedValue|,
+          [=string=] |languageTag:string|, and [=ordered map=]
+          |manifest:ordered map|:
+        </p>
+        <ol class="algorithm">
+          <li>Let |normalizedValue| be an [=ordered map=].
+          </li>
+          <li>If |localizedValue| is an [=ordered map=], and "value"
+          [=map/exists=] in |localizedValue|:
+            <ol>
+              <li>[=map/Set=] |normalizedValue|["value"] to
+              |localizedValue|["value"].
+              </li>
+              <li>If "lang" [=map/exists=] in |localizedValue|, [=map/set=]
+              |normalizedValue|["lang"] to |localizedValue|["lang"].
+              </li>
+              <li>If "dir" [=map/exists=] in |localizedValue|, [=map/set=]
+              |normalizedValue|["dir"] to |localizedValue|["dir"].
+              </li>
+            </ol>
+          </li>
+          <li>If "value" does not exist in |normalizedValue|, [=map/set=]
+          result["value"] to |localizedValue|.
+          </li>
+          <li>If "lang" does not exist in |normalizedValue|, [=map/set=]
+          |result|["lang"] to |languageTag|.
+          </li>
+          <li>If "dir" does not exist in |normalizedValue|, [=map/set=]
+          |normalizedValue|["dir"] to |manifest|["dir"].
+          </li>
+          <li>Return |normalizedValue|.
+          </li>
+        </ol>
+        <p>
+          To <dfn>process a localizable member</dfn>, given [=string=]
+          |localizableMemberName| and [=ordered map=] |json|:
+        </p><!-- todo: use algorithm when processing manifest -->
+        <ol class="algorithm">
+          <li>Let |localizedMemberName| be the [=string/concatenation=] of the
+          [=list=] « |localizableMemberName|, "_localized" ».
+          </li>
+          <li>If |localizedMemberName| [=map/exists=] in |json|:
+            <ol>
+              <li>Let |localizedMember| be |json|[|localizedMemberName|].
+              </li>
+              <li>If |localizedMember| is not an [=ordered map=], continue.
+              </li><!-- todo: any existing algorithm for this? -->
+              <!-- todo: also take manifest's own lang into account -->
```

Comment:
What's wanted here is `Intl.LocaleMatcher`, which is not yet beyond the proposal stage.

Language tag matching is complicated. It is case _insensitive_ and there can be some quirks if one wants the "best fit" style of matching. If one is okay with BCP47 Lookup matching, things are somewhat simpler: one selects the _longest_ tag that is a prefix of the user's requested value (or the default if no tag is a prefix). BCP47 [describes this](https://www.rfc-editor.org/rfc/rfc4647.html#section-3.4) in terms of truncating the requested tag until a match is found (or use the default if no match is found):

```
   Range to match: zh-Hant-CN-x-private1-private2
   1. zh-Hant-CN-x-private1-private2
   2. zh-Hant-CN-x-private1
   3. zh-Hant-CN
   4. zh-Hant
   5. zh
   6. (default)
```

Note that the manifest's own language should not participate in this. It has a language associated with it, but it is the "root locale" default value for purposes of language negotiation.

This could be a bit weird, since the localized value might override the manifest's default language (e.g. the default language is never visible)

Reply:
Resolution: Locale matching is left to the OS (e.g., needed for language change during runtime).